### ISSUE TITLE
Design of MPS class

### ISSUE THREAD

**Issue by stavros11** [google/TensorNetwork#205]
In #162 it was mentioned that it would great to add some built-in functionality for physics applications starting with MPS and DMRG. This could attract more researchers working on such topics. 

I am opening this to discuss about the MPS class. If nobody is already working on this I can contribute to the code part, but as I am not an MPS expert I would like some input regarding the class structure. Perhaps agreeing to a list of the methods we would like to have implemented would be a good start (Ash already made some recommendations in #155 and #162 ), 
We may even want to think for this in parallel with the QuantumTensorNetwork (start from a more general quantum state class?). It would also be great to be able to use the same MPS class for ML applications (tensor trains for the ML community).

The code under `experiments` could be useful to start, but I think we should set a requirement to keep this backend independent if possible.

**Comment by MichaelMarien** [comment]
We definitely should keep it as native to the entire TensorNetwork framework as possible and most functionality should be easily switchable between backends.

Few years ago, I dare say I was an expert in uniform MPS and for sure I can help with the required methods, class structure etc there. In fact, last weekend I was playing around a bit myself to port some of old Matlab code (https://github.com/MichaelMarien/VUMPS, I got stuck quite fast though).
I'm more than willing to contribute to this.

For not-uniform MPS, I never did much coding or research into this, so I can't say I know the pitfalls and crucial decisions along the way. Best to rely on some more experienced voices. Nevertheless, I for sure want to be of help!

I'm just not sure how best to start with this. I guess a lot people have very similar code bases lying around, and have some opinions about conventions and structure. Feels like a PR that one cannot just do on his own. At least a lot of discussion is needed, but I presume this might be al longer-living branch on which several people contribute. What's your preferred way of working on this?

**Comment by mganahl** [comment]
I have started to look into this, and the design I'm aiming at is to derive MPS from TensorNetwork. 
That has many advantages, can be run on different hardware easily, unified framework for U(N) and SU(N) symmetric tensors should be easily achievable, ... I have extensive experience with finite and infinite MPS optimization algorithms (with and without U(N) symmetries), and a design choice that works well for me (this is essentially the one Ash and I have implemented in experiments) is the following:

```python
class MPS(TensorNetwork):
  """
  A base class from which finite and infinite MPS are derived
  essentially, this is an infinite MPS with a finite unit-cell that implements most of
  the API. Examples of things that need to be implemented here below (I'm only listing a few here,
  I think the current API in experiments is a good starting point for a more complete list)
  I like this class to be typically in central gauge, i.e. have a center site (either bond or site). 
  """
  def transfer_operator(*args):
    """
    evolves a reduced density by one site
    """
   def transfer_matrix_eigs():
     """
     sparse eigenvalue decomposition of unit-cell transfer-operator
      """
    def measurement_related_things():
      "calls subclass methods"
    def regauging_related_things():
      "calls sub-class methods"
class InfiniteMPS(MPS):
  "An infinite MPS with finite unitcell; uMPS is a special case of that"

class FiniteMPS(MPS):
  "A finite MPS which has left and right most ancillary dimension 1." 

class PeriodicMPS(MPS):
  "a periodic MPS"
```
An `MPS` object has a finite unitcell, and different `MPS` object should be "stackable", i.e.
it should be easily possible to build a larger (e.g. two-unit-cell mps) from two seperate mps 
objects (they should gauge-connect smoothly). There are several approaches to that, and we should DEFINITELY have a few hangouts sessions where we discuss what's the best way of implementing this (I don't think there is one perfect way of doing this). As a starting point, I can explain everyone how I do this (it's also done this way in the current experiments implementation), but I'm happy to do something else if it makes sense. 

One thing we absolutely have to keep in mind is the extension to U(N) and SU(N) 
symmetric tensor networks (could be done by e.g. adding a new backend). We don't need to implement symmetries right away, but if we want to do so at some point, this affects some design choices. 

I think it should be fairly easy to take the current `MPS` implementation and restructure it so that it derives from TensorNetwork (we probably can throw out some classes there, or pull them together).

**Comment by Thenerdstation** [comment]
Hangout session to discuss this sounds like a great idea! We've been discussing having a non-internal group meeting with our new users and this seems like a great first topic.

**Comment by stavros11** [comment]
Sorry for the late response. I am not sure what's the best way to work on this, I just opened the issue to start a public discussion. Exactly because this is more extensive than a single PR, it would be good to have a place where everyone can get up-to-date with work in progress and potential users can comment on the API.

Since @mganahl wrote the existing MPS code and started looking into this, he probably has the most experience on how the coding should go. The meeting is definitely a good idea for more people to get involved.

Personally I have been looking at the existing code in `experiments/MPS/matrixproductstates.py` and made [a simple sketch of the structure and methods on Google Drawings](https://docs.google.com/drawings/d/1LxgQuHDu3jXkERo_pe276RJFimE5VOo2BBKU3rUfvgc/edit?usp=sharing) (perhaps not the best choice!). I share this here as it may help other people to get an idea of the current structure without looking at the raw code. Feel free to comment and edit this. Since MPS seems more complicated than `TensorNetwork` itself, such drawings may also help with coding or later for adding documentation. I did this by hand and will try to update this if there is some progress on new code (but note that there are automated tools that may give better results).
